pkg/apis/agones/v1/gameserver.go

```diff
@@ -926,3 +926,95 @@ func (gs *GameServer) Patch(delta *GameServer) ([]byte, error) {
 	result, err = json.Marshal(patch)
 	return result, errors.Wrapf(err, "error creating json for patch for GameServer %s", gs.ObjectMeta.Name)
 }
+
+// UpdateCount increments or decrements a CounterStatus on a Game Server by the given amount.
+func (gs *GameServer) UpdateCount(name string, action string, amount int64) error {
+	err := fmt.Errorf("unable to UpdateCount: Name %s, Action %s, Amount %d", name, action, amount)
+	if !(action == GameServerAllocationIncrement || action == GameServerAllocationDecrement) || amount < 0 {
+		return err
+	}
+	if counter, ok := gs.Status.Counters[name]; ok {
+		cnt := counter.Count
+		if action == GameServerAllocationIncrement {
+			cnt += amount
+			// only check for Count > Capacity when incrementing
+			if cnt > counter.Capacity {
+				return err
+			}
+		}
+		if action == GameServerAllocationDecrement {
+			cnt -= amount
+		}
+		if cnt < 0 {
+			return err
+		}
+		counter.Count = cnt
+		gs.Status.Counters[name] = counter
+		return nil
+	}
+	return err
+}
+
+// UpdateCounterCapacity updates the CounterStatus Capacity to the given capacity.
+func (gs *GameServer) UpdateCounterCapacity(name string, capacity int64) error {
+	err := fmt.Errorf("unable to UpdateCounterCapacity: Name %s, Capacity %d", name, capacity)
+	if capacity < 0 {
+		return err
+	}
+	if counter, ok := gs.Status.Counters[name]; ok {
+		counter.Capacity = capacity
+		gs.Status.Counters[name] = counter
+		return nil
+	}
+	return err
+}
+
+// UpdateListCapacity updates the ListStatus Capacity to the given capacity.
+func (gs *GameServer) UpdateListCapacity(name string, capacity int64) error {
+	err := fmt.Errorf("unable to UpdateListCapacity: Name %s, Capacity %d", name, capacity)
+	if capacity < 0 || capacity > 1000 {
+		return err
+	}
+	if list, ok := gs.Status.Lists[name]; ok {
+		list.Capacity = capacity
+		gs.Status.Lists[name] = list
+		return nil
+	}
+	return err
+}
+
+// AppendListValues adds given the values to the ListStatus Values list. Any duplicates are silently
+// dropped.
+func (gs *GameServer) AppendListValues(name string, values []string) error {
+	err := fmt.Errorf("unable to AppendListValues: Name %s, Values %s", name, values)
+	if len(values) == 0 {
+		return err
+	}
+	if list, ok := gs.Status.Lists[name]; ok {
+		vals := []string{}
+		// Maintain ordering so new values are appended to the end of list.Values
+		vals = append(vals, list.Values...)
+		vals = append(vals, values...)
+		vals = removeDuplicates(vals)
```

Comment:
I'll throw another potential idea at you, to see if it sticks.

Is it better to add then check if there are duplicates - or is it better to check to see if the value exists in the list, and if it does, don't add it again?

Looking at the code, it's a single loop either way (unless we want to keep a list + map for order + lookup speed) - but at this point it's probably just a micro optimisation, in which case ignore me 😄

Reply:
Can do. Made mergeRemoveDuplicates instead of removeDuplicates.